### PR TITLE
feat: add records/list to httpmux

### PR DESCRIPTION
**Pull request by daviddias**
Does: https://github.com/libp2p/hydra-booster/issues/27

WIP. Next:
- [x] Create endpoint
- [x] Test without records
- [ ] ~~Test with records~~ Done in a separate PR once ProviderManager is exposed

### CODE REVIEW

File: httpapi/httpapi.go [hunk 1]
```diff
@@ -23,9 +29,21 @@ func NewServeMux(nodes []*node.HydraNode) *http.ServeMux {
...
+func recordFetchHandler(nodes []*node.HydraNode) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// TODO
+	}
+}
+
+// "/records/list" Receive a record and fetch it from the network, if available
+func recordListHandler(nodes []*node.HydraNode) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// TODO
```
Comment (daviddias) [inline httpapi/httpapi.go]: Looking how to do this one out https://discuss.libp2p.io/t/list-provider-records/450

File: httpapi/httpapi.go [hunk 2]
```diff
@@ -23,9 +29,21 @@ func NewServeMux(nodes []*node.HydraNode) *http.ServeMux {
...
 		}
-	})
+	}
+}
 
-	return mux
+// "/records/fetch" Receive a record and fetch it from the network, if available
+func recordFetchHandler(nodes []*node.HydraNode) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// TODO
```
Comment (daviddias) [inline httpapi/httpapi.go]: Looking to see if my guess is correct https://discuss.libp2p.io/t/does-a-findproviders-replicate-the-provider-records-to-the-node-issuing-the-query/452

**Comment by daviddias** [comment]
To test with records will need to do https://github.com/libp2p/go-libp2p-kad-dht/issues/487 first.

@alanshaw mind to get this merged as is and do the last test in a separate PR? I'm afraid I've exhausted my hydra-booster dev time for today.

File: httpapi/httpapi_test.go [hunk 3]
```diff
@@ -58,3 +61,45 @@ func TestHTTPAPISybils(t *testing.T) {
 		}
 	}
 }
+
+func TestHTTPAPIRecordsListWithoutRecords(t *testing.T) {
+	ds := datastore.NewMapDatastore()
+	nodes, err := hytesting.SpawnNodes(1, []opts.Option{
+		opts.Datastore(ds),
+	}...)
```
Comment (alanshaw) [inline httpapi/httpapi_test.go]:
```suggestion
	nodes, err := hytesting.SpawnNodes(1, opts.Datastore(ds))
```

File: httpapi/httpapi_test.go [hunk 4]
```diff
@@ -58,3 +61,45 @@ func TestHTTPAPISybils(t *testing.T) {
 		}
 	}
 }
+
+func TestHTTPAPIRecordsListWithoutRecords(t *testing.T) {
+	ds := datastore.NewMapDatastore()
```
Comment (alanshaw) [inline httpapi/httpapi_test.go]: Can we put a fake entry directly in the datastore to retrieve?

File: httpapi/httpapi.go [hunk 5]
```diff
@@ -23,16 +41,39 @@ func NewServeMux(nodes []*node.HydraNode) *http.ServeMux {
 				Addrs: n.Host.Addrs(),
 			})
 		}
-	})
+	}
+}
 
-	return mux
+// "/records/fetch" Receive a record and fetch it from the network, if available
```
Comment (alanshaw) [inline httpapi/httpapi.go]: 🙏 please can we have docs for these new endpoints in the README?

File: httpapi/httpapi.go [hunk 6]
```diff
@@ -23,16 +41,39 @@ func NewServeMux(nodes []*node.HydraNode) *http.ServeMux {
...
+		// https://discuss.libp2p.io/t/list-provider-records/450
+		// for now, enumerate the Provider Records in the datastore
+
+		providersKeyPrefix := "/providers/" // https://github.com/libp2p/go-libp2p-kad-dht/blob/master/providers/providers.go#L76
+		results, err := datastore.Query(dsq.Query{
+			Prefix: providersKeyPrefix,
+		})
+		if err != nil {
+			w.WriteHeader(500)
+			return
```
Comment (alanshaw) [inline httpapi/httpapi.go]: Please log out the error!

File: httpapi/httpapi.go [hunk 7]
```diff
@@ -23,16 +41,39 @@ func NewServeMux(nodes []*node.HydraNode) *http.ServeMux {
...
-		return err
+// "/records/list" Receive a record and fetch it from the network, if available
+func recordListHandler(nodes []*node.HydraNode, datastore ds.Batching) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// TODO Improve this handler once ProvideManager gets exposed
+		// https://discuss.libp2p.io/t/list-provider-records/450
+		// for now, enumerate the Provider Records in the datastore
+
+		providersKeyPrefix := "/providers/" // https://github.com/libp2p/go-libp2p-kad-dht/blob/master/providers/providers.go#L76
+		results, err := datastore.Query(dsq.Query{
```
Comment (alanshaw) [inline httpapi/httpapi.go]: I know this is weird, but we know they share the same datastore...

```suggestion
		results, err := nodes[0].Datastore.Query(dsq.Query{
```

Do you think we need a:

```go
type Hydra struct {
	Heads		[]*head.Head // or maybe `Sybils`?
	Datastore	datastore.Datastore
	RoutingTable	*kbucket.RoutingTable
}
```

? Where `head.Head` is basically what `HydraNode` is now:

```go
type Head struct {
	Host         host.Host
	Routing      routing.Routing
	Bootstrapped bool
}
```

...and we'd pass our `Hydra` instance to `NewServeMux`.

Maybe one for a separate PR?

Comment (daviddias) [reply httpapi/httpapi.go]: I think that is a premature refactor. Right now, the cost is just one repeated pointer.

File: httpapi/httpapi.go (hunk 6 above)
Comment (daviddias) [reply httpapi/httpapi.go]: Of course!!

File: httpapi/httpapi.go (hunk 5 above)
Comment (daviddias) [reply httpapi/httpapi.go]: Of course! :)

File: httpapi/httpapi_test.go (hunk 4 above)
Comment (daviddias) [reply httpapi/httpapi_test.go]: @alanshaw once we have the ProviderManager, then we quickly can do a pm.AddProvider. Was waiting for that, sounds good?

Comment (alanshaw) [reply httpapi/httpapi_test.go]: Ok sure!

**Comment by daviddias** [comment]
Applied CR, merging, now onto the ProviderManager thing

**Pull request merged**